permissioned-token/program/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+    )?;
+    Ok(seeds)
+}
+
+#[inline]
+fn get_freeze_authority(upstream_authority: &Pubkey) -> (Pubkey, Vec<Vec<u8>>) {
+    let mut seeds = vec![upstream_authority.as_ref().to_vec()];
+    let (freeze_key, bump) = Pubkey::find_program_address(
+        &seeds.iter().map(|s| s.as_slice()).collect::<Vec<&[u8]>>(),
+        &crate::id(),
+    );
+    seeds.push(vec![bump]);
+    (freeze_key, seeds)
+}
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    let instruction = PermissionedTokenInstruction::try_from_slice(instruction_data)?;
+    match instruction {
+        PermissionedTokenInstruction::InitializeMint { decimals } => {
+            msg!("PermissionedTokenInstruction::InitializeMint");
+            process_initialize_mint(program_id, accounts, decimals)
+        }
+        PermissionedTokenInstruction::InitializeAccount => {
+            msg!("PermissionedTokenInstruction::InitializeAccount");
+            process_initialize_account(accounts)
+        }
+        PermissionedTokenInstruction::Transfer { amount } => {
+            msg!("PermissionedTokenInstruction::Transfer");
+            process_transfer(accounts, amount)
+        }
+        PermissionedTokenInstruction::MintTo { amount } => {
+            msg!("PermissionedTokenInstruction::MintTo");
+            process_mint_to(program_id, accounts, amount)
+        }
+        PermissionedTokenInstruction::Burn { amount } => {
+            msg!("PermissionedTokenInstruction::Burn");
+            process_burn(accounts, amount)
+        }
+        PermissionedTokenInstruction::CloseAccount => {
+            msg!("PermissionedTokenInstruction::CloseAccount");
+            process_close(accounts)
+        }
+    }
+}
+
+pub fn process_initialize_mint(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    decimals: u8,
+) -> ProgramResult {
+    let InitializeMint {
+        mint,
+        payer,
+        upstream_authority,
+        system_program,
+        token_program,
+    } = InitializeMint::load(accounts)?;
+    let space = spl_token::state::Mint::LEN;
+    invoke(
+        &system_instruction::create_account(
+            payer.key,
+            mint.key,
+            Rent::get()?.minimum_balance(space),
+            space as u64,
+            token_program.key,
+        ),
+        &[payer.clone(), mint.clone(), system_program.clone()],
+    )?;
+    let (mint_authority, _) = Pubkey::find_program_address(&[mint.key.as_ref()], program_id);
+    let (freeze_key, _) = get_freeze_authority(upstream_authority.key);
+    initialize_mint(&freeze_key, &mint_authority, mint, token_program, decimals)
+}
+
+pub fn process_initialize_account(accounts: &[AccountInfo]) -> ProgramResult {
+    let InitializeAccount {
+        token_account,
+        owner,
+        payer,
+        upstream_authority,
+        freeze_authority,
+        mint,
+        system_program,
+        rent,
```

Comment:
You can remove the rent account, it isn't needed in the ATA program!

Reply:
Is this a recent change? I feel like it used to be needed

Reply:
I think it was needed for ATA version 1.0.4 and before